lib/p2p/packets/packetUtils.ts

```diff
@@ -1,30 +1,52 @@
 import PacketType from './PacketType';
-import Packet from './Packet';
+import Packet, { PacketInterface } from './Packet';
 import * as packetTypes from './types';
 
-const fromRaw = (type: string, packet: string): Packet | void => {
-  switch (type) {
-    case PacketType.HELLO:
-      return new packetTypes.HelloPacket(packet);
-    case PacketType.PING:
-      return new packetTypes.PingPacket(packet);
-    case PacketType.PONG:
-      return new packetTypes.PongPacket(packet);
-    case PacketType.ORDER:
-      return new packetTypes.OrderPacket(packet);
-    case PacketType.GET_ORDERS:
-      return new packetTypes.GetOrdersPacket(packet);
-    case PacketType.ORDERS:
-      return new packetTypes.OrdersPacket(packet);
-    case PacketType.GET_HOSTS:
-      return new packetTypes.GetHostsPacket(packet);
-    case PacketType.HOSTS:
-      return new packetTypes.HostsPacket(packet);
-    default:
-      return;
+class ParserError {
+  constructor(public type: ParserErrorType, public payload: string) { }
+}
+
+enum ParserErrorType {
+  UNPARSABLE_MESSAGE,
+  INVALID_MESSAGE,
+  UNKNOWN_PACKET_TYPE,
+}
+
+const fromRaw = (raw: string): Packet => {
+  try {
+    const json = JSON.parse(raw);
+    // check that we have the required fields for an incoming packet
+    if (typeof json.header === 'object' && typeof json.header.type === 'string' && typeof json.header.id === 'number') {
+      const packet = json as PacketInterface;
+      switch (packet.header.type) {
+        case PacketType.HELLO:
+          return new packetTypes.HelloPacket(packet);
+        case PacketType.PING:
+          return new packetTypes.PingPacket(packet);
+        case PacketType.PONG:
+          return new packetTypes.PongPacket(packet);
+        case PacketType.ORDER:
+          return new packetTypes.OrderPacket(packet);
+        case PacketType.GET_ORDERS:
+          return new packetTypes.GetOrdersPacket(packet);
+        case PacketType.ORDERS:
+          return new packetTypes.OrdersPacket(packet);
+        case PacketType.GET_HOSTS:
+          return new packetTypes.GetHostsPacket(packet);
+        case PacketType.HOSTS:
+          return new packetTypes.HostsPacket(packet);
+        default:
+          throw new ParserError(ParserErrorType.UNKNOWN_PACKET_TYPE, packet.header.type!);
```

Comment:
this will be catch and by the outer `try`/`catch`, so the error type will be misleading.